src/wasm/wasm-type.cpp

```diff
@@ -1277,27 +1280,86 @@ void Canonicalizer::scanType(Type* type) {
   }
 }
 
-void Canonicalizer::makeReachabilityFixedPoint() {
+void Canonicalizer::makeAncestorsFixedPoint() {
   // Naively calculate the transitive closure of the reachability graph.
   bool changed;
   do {
     changed = false;
-    for (auto& entry : reaches) {
-      auto& reachable = entry.second;
-      std::unordered_set<TypeID> nextReachable;
-      for (auto& other : reachable) {
-        auto& otherReaches = reaches[other];
-        nextReachable.insert(otherReaches.begin(), otherReaches.end());
+    for (auto& entry : ancestors) {
+      auto& succs = entry.second;
+      std::unordered_set<TypeID> nextAncestors;
+      for (auto& other : succs) {
+        auto& otherAncestors = ancestors[other];
+        nextAncestors.insert(otherAncestors.begin(), otherAncestors.end());
       }
-      size_t oldSize = reachable.size();
-      reachable.insert(nextReachable.begin(), nextReachable.end());
-      if (reachable.size() != oldSize) {
+      size_t oldSize = succs.size();
+      succs.insert(nextAncestors.begin(), nextAncestors.end());
+      if (succs.size() != oldSize) {
         changed = true;
       }
     }
   } while (changed);
 }
 
+std::vector<Canonicalizer::Item> Canonicalizer::getOrderedItems() {
+  // Topological sort the types and heaptypes so that all children are
+  // canonicalized before their parents.
+
+  std::vector<TypeID> sorted;
+  std::unordered_set<TypeID> seen;
+
+  std::function<void(TypeID)> visit = [&](TypeID i) {
+    if (seen.count(i)) {
+      return;
+    }
+    // Visit all children before the current type.
+    auto childrenIt = ancestors.find(i);
+    if (childrenIt != ancestors.end()) {
+      for (auto child : childrenIt->second) {
+        visit(child);
```

Comment:
I'm missing something. `ancestors` is defined as
```cpp
  // Maps Type and HeapType IDs to the IDs of their ancestor Types and HeapTypes
  // in the type graph. Only considers compound Types and HeapTypes.
```
From this description I expect `ancestors[id]` to be a list of the ancestors of `id`. But the loop here instead calls those items "children" - ?

Reply:
Oh yeah, I got myself confused about the correct direction of the relation when I was writing this and I guess I forgot to update this code. I'll fix this. `ancestors` and your expectation are correct.

Reply:
@kripken this should be fixed now.